OceanLab/eof.py

```diff
@@ -121,3 +121,162 @@ def my_eof_interp(M,nmodes,errmin=1e-15,repmax=None):
+        New modifications:
+        (i)  The function organizes the data as time vs space and subtract the mean field in each coordinate
+        (ii) It returns all the variables related to CEOF in a DataArray
+    ''' 
+    # Organizing the data as time vs space
+    data_ceof = org_data_ceof(lon, lat, data)
+    # We need to remove the mean field (i.e., the trend) in each coordinate to 
+    # evaluate the variability 
+    data_ceof = data_ceof - np.nanmean(data_ceof,axis=0)
+    
+    # The variables below are useful later
+    load_real = np.zeros([data_ceof.shape[1], nkp])*np.nan
+    load_imag = np.zeros([data_ceof.shape[1], nkp])*np.nan
+    # It is necessary to remove the nan values of the matrix to solve the eigenvalue problem
+    nan_values = np.isnan(data_ceof[0,:]) # We can just look at each coordinate along a single time
+    data_ceof = data_ceof[:,~nan_values]   # Then, we remove all these coordinates in all of the occurences
+    
+    ntim, npt = data_ceof.shape
+    
+    # Hilbert transform: input sequence x and returns a complex result of the same length
+    print('1: Performing Hilbert transform')
+    data_hilbert = hilbert(data_ceof)
+    # Compute the covariance matrix in the Hilbert transform
+    print('2: Computing covariance matrix')
+    c = delayed(np.dot)(data_hilbert.conjugate().T, data_hilbert).compute()/ntim
+    print('3: Solving the eigenvalue problem')
+    lamda, loadings = delayed(la.eig)(c).compute() # lamda: eigenvalue, loadings: eigenvectors
+    
+    l = lamda.conjugate().T; k = np.argsort(l)
+    lamda, loadings = np.flip(l[k]), np.fliplr(loadings[:,k])
+    loadings = loadings[:,:nkp]
+    # In case there were nan values in the orginal data, we need to perform the approach below:
+    load_real[~nan_values,:] = loadings.real.copy()
+    load_imag[~nan_values,:] = loadings.imag.copy()
+    load = load_real + 1j*load_imag
+    modes = load.reshape((len(lat),len(lon), nkp))
+    
+    per = lamda.real*100/np.sum(lamda.real)
+    per = per[:nkp].copy()
+    pcs = np.dot(data_hilbert,loadings)
+    
+    sp_amp, sp_phase, t_amp, t_phase = amplitude_phase(load, pcs)
+    sp_amp   = sp_amp.reshape((len(lat),len(lon), nkp))
+    sp_phase = sp_phase.reshape((len(lat),len(lon), nkp))    
+    
+    print('Done! \U0001F600')
+    
+    dims = ["lat", "lon", "nkp", "time"]
+    ds = xr.Dataset({"per":(dims[2], per),"modes":(dims[:-1], modes),"sp_amp":(dims[:-1], sp_amp),
+                    "sp_phase":(dims[:-1], sp_phase),"pcs":(dims[-2:][::-1], pcs),"t_amp":(dims[-2:][::-1], t_amp),
+                    "t_phase":(dims[-2:][::-1], t_phase)},
+                    coords={"lat":(dims[0], lat), "lon":(dims[1], lon), "nkp":(dims[2], np.arange(nkp)),
+                           "time":(dims[3], np.arange(len(data_ceof)))})
+
+    return ds
+
+def org_data_ceof(lon, lat, data):
+    data_ceof = np.zeros((len(data), len(lon)*len(lat)))*np.nan
+    k = 0
+    for i in range(len(lat)):
```

Comment:
This method needs to be refactored. The computation time of a nested loop increases exponentially as the list size increases.